src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,142 @@
+"""
+Perform Chavatal-Gomory strengthening of inequalities
+
+We apply the Chvatal-Goromory procedure to constraints that
+satisfy the following criterions:
+1. All variables with nonzero coefficient must be integral
+2. The constraints are not ranged
+3. If the constraints is of the form ax ⩾ b, then every
+   positive coefficient variables must have finite lowerbound
+   negative coefficient variables must have finite upperbound
+
+   If the constraints is of the from ax ⩽ b, then every
+   positive coefficient variables must have finite upperbound
+   negative coefficient variables must have finite lowerbound
+
+For example, let aᵢx ⩾ bᵢ is a constraint where aᵢ ⩾ 0, and x
+is integer variables. If there exist s > 0 such that
+⌈aᵢⱼ s⌉ bᵢ / ⌈bᵢ s⌉ ⩽ aᵢⱼ
+and there is k such that
+⌈aᵢₖ s⌉ bᵢ / ⌈bᵢ s⌉ < aᵢₖ (*)
+then we will update aᵢx ⩾ bᵢ into ⌈aᵢ s⌉ x ⩾ ⌈bᵢ s⌉
+
+We find such a number s heuristicaly i.e, we try different
+values from the following set
+{1, t/a_max, t/a_min, (2t-1)/(2a_min)| t = 1,...,search_range}
+where a_max = sup(|aᵢ|), a_min = inf(|aᵢ|) and search_range
+is set to be 5 by default
+
+"""
+function cg_strengthening!(ps::PresolveData{T}, search_range::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+
+    i_index = zeros(Int, ps.pb0.nvar)
+    for i in 1:ps.pb0.ncon
```

Comment:
I think this should be `nrow` not `ncon`

Reply:
It would be good to rewrite one of your tests such that it fails without this fix.

Reply:
ps.nrow and ps.pb0.ncon are the same thing (line 112 presolve.jl), but I think I will change to ps.nrow.

However,I think another test is not needed. In one of the tests, I have a ranged constraints, if ncon counted twice for that one, it should have been an out-of-index error

Reply:
That's in the constructor. Is that invariant guaranteed after, e.g., other reductions have been applied?

Reply:
yes, you're right. I'll fix that and add a test